crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs

```diff
@@ -0,0 +1,132 @@
+use ruff_diagnostics::{Diagnostic, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::Stmt;
+use ruff_python_semantic::Binding;
+
+use crate::{
+    checkers::ast::Checker,
+    renamer::{try_shadowed_kind, Renamer, ShadowedKind},
+};
+
+/// ## What it does
+///
+/// Checks for use of [PEP 695] type parameters with leading underscores in generic classes and
+/// functions.
+///
+/// ## Why is this bad?
+///
+/// [PEP 695] type parameters are already restricted in scope to the class or function in which they
+/// appear, so leading underscores just hurt readability without the usual privacy benefits.
+///
+/// ## Example
+///
+/// ```python
+/// class GenericClass[_T]:
+///     var: _T
+///
+///
+/// def generic_function[_T](var: _T) -> list[_T]:
+///     return var[0]
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// class GenericClass[T]:
+///     var: T
+///
+///
+/// def generic_function[T](var: T) -> list[T]:
+///     return var[0]
+/// ```
+///
+/// ## Fix availability
+///
+/// This rule avoids shadowing builtins, keywords, and other variables with new type parameter names
+/// and will offer a diagnostic but not a fix in the case of a conflict.
```

Comment:
```suggestion
/// If the name without an underscore would shadow a builtin or another variable,
/// would be a keyword, or would otherwise be an invalid identifier, a fix will not
/// be available. In these situations, you can consider using a trailing underscore
/// or a different name entirely to satisfy the lint rule.
```